tests/extra_source_and_tests/test_2d_cohesive_soil_failure/cohesive_soil_failure.h

```diff
@@ -0,0 +1,307 @@
+Real gravity_g = 9.8;                                                     // gravity force of soil
+Real Youngs_modulus = 1.8e6;                                              // reference Youngs modulus
+Real poisson = 0.3;                                                       // Poisson ratio
+Real c_s = sqrt(Youngs_modulus / (rho0_s * 3.0 * (1.0 - 2.0 * poisson))); // sound speed
+Real cohesion = 5.0e3;
+Real friction_angle = 25.0 * Pi / 180;
+//----------------------------------------------------------------------
+//	Geometric shapes used in this case.
+//----------------------------------------------------------------------
+Vec2d soil_block_halfsize = Vec2d(0.5 * LL, 0.5 * LH); // local center at origin:
+Vec2d soil_block_translation = soil_block_halfsize;
+Vec2d outer_wall_halfsize = Vec2d(0.5 * DL + BW, 0.5 * DH + BW);
+Vec2d outer_wall_translation = Vec2d(-BW, -BW) + outer_wall_halfsize;
+Vec2d inner_wall_halfsize = Vec2d(0.5 * DL, 0.5 * DH);
+Vec2d inner_wall_translation = inner_wall_halfsize;
+//----------------------------------------------------------------------
+//	Complex for wall boundary
+//----------------------------------------------------------------------
+class WallBoundary : public ComplexShape
+{
+  public:
+    explicit WallBoundary(const std::string &shape_name) : ComplexShape(shape_name)
+    {
+        add<TransformShape<GeometricShapeBox>>(Transform(outer_wall_translation), outer_wall_halfsize);
+        subtract<TransformShape<GeometricShapeBox>>(Transform(inner_wall_translation), inner_wall_halfsize);
+    }
+};
+std::vector<Vecd> soil_shape{
+    Vecd(0, 0), Vecd(0, LH), Vecd(LL, LH), Vecd(LL, 0), Vecd(0, 0)};
+
+class Soil : public MultiPolygonShape
+{
+  public:
+    explicit Soil(const std::string &shape_name) : MultiPolygonShape(shape_name)
+    {
+        multi_polygon_.addAPolygon(soil_shape, ShapeBooleanOps::add);
+    }
+};
+//----------------------------------------------------------------------
+//	application dependent initial condition
+//----------------------------------------------------------------------
+class SoilInitialCondition : public continuum_dynamics::ContinuumInitialCondition
+{
+  public:
+    explicit SoilInitialCondition(RealBody &granular_column)
+        : continuum_dynamics::ContinuumInitialCondition(granular_column){};
+
+  protected:
+    void update(size_t index_i, Real dt)
+    {
+        /** initial stress */
+        Real y = pos_[index_i][1];
+        Real gama = 1 - sin(friction_angle);
+        Real stress_yy = -rho0_s * gravity_g * y;
+        stress_tensor_3D_[index_i](1, 1) = stress_yy;
+        stress_tensor_3D_[index_i](0, 0) = stress_yy * gama;
+        stress_tensor_3D_[index_i](2, 2) = stress_yy * gama;
+    };
+};
+//----------------------------------------------------------------------
+//	Unified transport velocity correction
+//----------------------------------------------------------------------
+template <typename... T>
```

Comment:
I am wondering why we need this new class. Could we use a generalization-extension approach to add this new feature?

Reply:
> I am wondering why we need this new class. Could we use a generalization-extension approach to add this new feature?

Yes. It's better to use a generalized approach.
A simple approach is to just add a judgment statement in the original `TransportVelocityCorrection`.
When a particle is identified as a surface particle, corrections are applied only in the tangential direction. However, it is uncertain whether this approach will affect other cases.

Reply:
No. Using a switch is not a generalization-extension approach. We need first find a type in the original formulation such that it can be generalized, then change it to a template parameter (as a templated class) and reimplement the original class by realizing the template parameter with a small function object. For extension, we write the new feature as a new function object which substitute the template parameter in the generalized templated class. In this way, the choose new feature is don in compiling time not runtime.

Reply:
> No. Using a switch is not a generalization-extension approach. We need first find a type in the original formulation such that it can be generalized, then change it to a template parameter (as a templated class) and reimplement the original class by realizing the template parameter with a small function object. For extension, we write the new feature as a new function object which substitute the template parameter in the generalized templated class. In this way, the choose new feature is don in compiling time not runtime.

I will think about this.